### DISCUSSION TITLE
Is the Prisma Client faster because of Rust?

### DISCUSSION THREAD

**Discussion by enricoschaaf** [prisma/prisma#2678] [Q&A] [answered]
I was wondering if there are any speed advantages, for example how many requests the Prisma Client can handle, that come with Rust compared to other libraries or languages?

**Comment by ryands17** [comment, marked as answer]
Hey @enricoschaaf :wave: 
A couple of reasons for implementing in Rust would be:
1. To implement Prisma for multiple languages. Currently Prisma is mainly focusing on Node, but there is a Go Client as well in progress. Using Rust means not depending on any runtime like Node, so we can expand this to more languages.
2. Rust does offer improved performance as it doesn't need any external runtime like Node or Python and is compiled directly. It also has a lower memory footprint which is also the reason Prisma 2 was based on Rust and not on it's precedecessor Prisma 1 which used Scala (JVM).

So if you're coming from Prisma 1, you would notice there's no need for an external Prisma server that was required because of JVM.

**Reply by marcj** [reply]
> Rust does offer improved performance

Sure, Rust is faster than Node. But the question was if Prisma Client is actually faster because of Rust? In our benchmarks, Prisma is pretty slow compared to other ORMs. In fact it was the by far slowest. Do you guys track the actual performance?

**Reply by nikolasburk** [reply]
Hey Marc 👋 we've just started an internal benchmarking effort and will use this to identify bottlenecks and optimize Prisma Client's performance. (We've been following a "make it work, make it right, make it fast" approach for building Prisma and just now arrive at the last stage of this).

I've also reached out to you via email (using the email address from your GitHub bio), we'd love to talk to you regarding your benchmarking efforts if you're open it.

**Reply by nikolasburk** [reply]
@marcj I just answered another performance-related question here: https://github.com/prisma/prisma/discussions/5819#discussioncomment-404842

TLDR: One major reason why Prisma Client currently might be slower than other ORMs is because of the overhead that's created when the Node.js language layer of Prisma Client communicates with the underlying query engine via HTTP (which is just due to pragmatic legacy architectural decisions that we've made in the past). However, we are currently [working towards replacing this interface](https://github.com/prisma/prisma/issues/4911) using the [N-API](https://napi.rs/) and are getting very close to having this implemented. This should notably speed up Prisma Client queries already 🙂 

By the way, congrats on the progress on Deepkit, really looking forward to the launch 🙌 😄

**Comment by ludralph** [comment]
Hi @enricoschaaf,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
